2-Draft-Accepted/RFCNNNN-Support-Experimental-Features.md

```diff
@@ -0,0 +1,626 @@
+  BeforeAll {
+    $skip = -not $PSEnabledExperimentalFeatures.Contains("X")
+    ...
+  }
+
+  It "..." -Skip:$skip {
+    ...
+  }
+}
+
+try {
+  $skip = -not $PSEnabledExperimentalFeatures.Contains("X")
+  $originalDefaultParameterValues = $PSDefaultParameterValues.Clone()
+  $PSDefaultParameterValues["it:skip"] = $skip
+
+  Describe "Tests for Experimental Feature X" {
+    ...
+  }
+}
+finally {
+  $global:PSDefaultParameterValues = $originalDefaultParameterValues
+}
+```
+
+### Experimental Feature Incompatibility Mitigation
+
+There will be cases where experimental features are incompatible either intentionally or unintentionally.
+
+- An example of an intentional incompatibility occurs when two alternative, but incompatible solutions, are being tested.
+- An example of an unintended incompatibility can occur when two experimental features prove to be incompatible through usage.
+
+A minimal implementation will support defining a list of experimental feature pairs in the `powershell.config.json` file.
+At startup, attempts to use incompatible features would produce warnings and consider both as disabled.
+
+An opportunistic implementation will support an exclusion value in the experimental feature declaration
+for cases where an author is intentionally providing alternative/incompatible implementations.
+
+### Static Code Analysis
+
+Static code analysis support should be provided to address experimental feature authoring.
+Two areas are particularly important:
+
+1. Validating whether an experimental feature is marked correctly.
+   A use case for this occurs when an experimental feature is referenced in a module but not, or incorrectly defined in the manifest.
+   From a script perspective, a `PSScriptAnalyzer` rule would be a reasonable solution.
+   From a compiled code perspective, a reflection-based tool may be needed.
+
+2. Validating whether references to an non-existing experimental feature have been removed.
+   At some point, the experiment is completed and either removed from or incorporated into the code base.
+   At this point, a tool to scan script or compiled code would be needed to ensure any references to the experiment have been removed.
+
+## Open Issues
+
+### Whether experimental features can be included in PowerShell GA releases
+
+It's arguable whether or not experimental features should be allowed in PowerShell GA releases.
+Take C# an an instance, it has experimental language features while in the development of a new version.
+However, when reaching the final release,
+there won't be any experimental features -- a feature is either officially kept or removed from the final release.
+
+The benefit of not allowing experimental features in a GA release is to improve stability.
+But the drawback is that we cannot get more usage and feedback for an experimental feature
+before deciding to keep or cut the feature.
+This is because preview releases of PowerShell usually get much less usage compared with the GA release.
```

Comment:
Only a small number of users need these features. Even if they will be in GA they will not use any more.
Since this is an open source project then anyone can even compile the product with the desired features. This dynamic enabling experimental features is only needed to make it easier for users to get these features without compiling. So I think it's better to have a special (debug/dev) build that supports these features. It can also provide a separate telemetry. Special (debug/dev) build is also needed if we worry about "to improve stability GA" (line 557). So Alpha/Beta/Preview/GA withoutexperimental features and Dev with experimental features.

Reply:
The only other language where I've seen experimental features really exposed with a switch is [Haskell](http://downloads.haskell.org/~ghc/latest/docs/html/users_guide/glasgow_exts.html) where they can be activated with either a compiler flag or a pragma. They've compiled [a few tips](https://wiki.haskell.org/Use_of_language_extensions) for using and implementing experimental features there. My own experience with them is that, since they are available in releases, people use them inconsistently and they remain experimental indefinitely -- Haskell essentially has what amount to minor dialects based on what extensions are enabled. The difference is that with us, experimental features are not necessarily just *compiler* features, but in the runtime. So allowing experimental feature switches in releases might balkanize PowerShell installations, and (perhaps more importantly) modules.